### ISSUE TITLE
DemoBench: key sequence hI is interpreted as up arrow key

### ISSUE THREAD

**Issue by mikehearn** [corda/corda#423]
This is an odd one. On macOS at least, typing "hI" into the shell in the terminal is interpreted as if you had pressed the up arrow key. Some sort of escaping/configuration malarky going on here. I noticed it when trying to type CashIssue

**Comment by chrisr3** [comment]
On Fedora/Linux, the key sequence "hI" triggers an NPE:
```
Sat Mar 25 20:44:23 GMT 2017>>> hException in thread "Command line shell processor" java.lang.NullPointerException
        at org.crsh.console.EditorBuffer.replace(EditorBuffer.java:171)
        at org.crsh.console.EditorAction$History.perform(EditorAction.java:224)
        at org.crsh.console.EditorAction.execute(EditorAction.java:768)
        at org.crsh.console.Editor.append(Editor.java:151)
        at org.crsh.console.Console.iterate(Console.java:214)
        at org.crsh.console.Console.on(Console.java:158)
        at org.crsh.console.Console.on(Console.java:135)
        at org.crsh.console.jline.JLineProcessor.run(JLineProcessor.java:204)
        at net.corda.node.InteractiveShell$startShell$2.invoke(InteractiveShell.kt:147)
        at net.corda.node.InteractiveShell$startShell$2.invoke(InteractiveShell.kt:66)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
```

**Comment by chrisr3** [comment]
The following keys have unexpected effects in CRaSH:

Key | Input
------ | --------
G | BEGINNING_OF_LINE
H | PREVIOUS_HISTORY
I | BEGINNING_OF_HISTORY
K | BACKWARD_CHAR
M | FORWARD_CHAR
O | END_OF_LINE
P | NEXT_HISTORY
Q | END_OF_HISTORY
R | OVERWRITE_MODE
S | DELETE_CHAR

I can replicate these results with JLine alone via:
```java
int i = 71; // 72, ..., 83
StringBuilder sb = new StringBuilder()
    .appendCodePoint(0)
    .appendCodePoint(i);

KeyMap keyMap = reader.getKeys();
Object obj = keyMap.getBound(sb);
```

**Comment by chrisr3** [comment]
These are bound as "arrow keys" by JLine:
```java
jline.console.KeyMap.bindArrowKeys(KeyMap map):

bind( map, "\000\107", Operation.BEGINNING_OF_LINE );
bind( map, "\000\110", Operation.PREVIOUS_HISTORY );
bind( map, "\000\111", Operation.BEGINNING_OF_HISTORY );
bind( map, "\000\110", Operation.PREVIOUS_HISTORY );
bind( map, "\000\113", Operation.BACKWARD_CHAR );
bind( map, "\000\115", Operation.FORWARD_CHAR );
bind( map, "\000\117", Operation.END_OF_LINE );
bind( map, "\000\120", Operation.NEXT_HISTORY );
bind( map, "\000\121", Operation.END_OF_HISTORY );
bind( map, "\000\122", Operation.OVERWRITE_MODE );
bind( map, "\000\123", Operation.DELETE_CHAR );
```

**Comment by mikehearn** [comment]
But why is a `h` being interpreted as a null?

**Comment by chrisr3** [comment]
AFAICT, the 'h' _is_ being handled correctly. Which is why I get a `hException` message :-).
```
Mon Mar 27 13:52:36 BST 2017>>> hException in thread "Command line shell processor" java.lang.NullPointerException
```

**Comment by chrisr3** [comment]
As a "proof of concept", I have rebuilt `jline` without those `bind()` statements (above), rebuilt `crash.shell` to contain the new `jline` (because it "shades" `jline` to make an internal copy) and then rebuilt Corda with the new `crash.shell`. At which point `G`, `H`, `I`, `K`, etc start behaving correctly.

**Comment by chrisr3** [comment]
I think the real problem here is that CRaSH is receiving keycode 0 (zero) whenever SHIFT, CTRL, ALT etc keys are pressed. If `JLineProcessor` never receives `\000` then it cannot map `\000\111` to `BEGINNING_OF_HISTORY`.

**Comment by chrisr3** [comment]
This is looking like a configuration issue with JediTerm. Given `getinput.jar`, which prints the keycodes returned from `System.in` until you press 'q`:
- Launching `getinput.jar` within `JediTermWidget` from IntelliJ works OK - no keycodes for SHIFT, CTRL are returned.
- Launching `getinput.jar` from DemoBench does _not_ work - keycode 0 returned for SHIFT, CTRL.
I am assuming that IntelliJ is performing some extra setup on the PTY that processes launched from it can inherit.

**Comment by chrisr3** [comment]
I have now reproduced the problem by embedding `JediTermWidget` within a test JavaFX application.

**Comment by chrisr3** [comment]
Fixed by #460.
